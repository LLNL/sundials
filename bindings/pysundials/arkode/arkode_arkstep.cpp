/*------------------------------------------------------------------------------
 * Programmer(s): Cody J. Balos @ LLNL
 *------------------------------------------------------------------------------
 * SUNDIALS Copyright Start
 * Copyright (c) 2025-2025, Lawrence Livermore National Security,
 * University of Maryland Baltimore County, and the SUNDIALS contributors.
 * Copyright (c) 2013-2025, Lawrence Livermore National Security
 * and Southern Methodist University.
 * Copyright (c) 2002-2013, Lawrence Livermore National Security.
 * All rights reserved.
 *
 * See the top-level LICENSE and NOTICE files for details.
 *
 * SPDX-License-Identifier: BSD-3-Clause
 * SUNDIALS Copyright End
 *----------------------------------------------------------------------------*/

#include <nanobind/nanobind.h>
#include <nanobind/ndarray.h>
#include <nanobind/stl/function.h>

#include <arkode/arkode.hpp>
#include <arkode/arkode_arkstep.h>
#include <sundials/sundials_core.hpp>

#include "arkode_usersupplied.hpp"

namespace nb = nanobind;

namespace pysundials {

void bind_arkode_arkstep(nb::module_& m)
{
#include "arkode_arkstep_generated.hpp"

  m.def(
    "ARKStepCreate",
    [](std::function<std::remove_pointer_t<ARKRhsFn>> fe,
       std::function<std::remove_pointer_t<ARKRhsFn>> fi, sunrealtype t0,
       N_Vector y0, SUNContext sunctx)
    {
      auto fe_wrapper = fe ? arkstep_fe_wrapper : nullptr;
      auto fi_wrapper = fi ? arkstep_fi_wrapper : nullptr;

      void* ark_mem = ARKStepCreate(fe_wrapper, fi_wrapper, t0, y0, sunctx);
      if (ark_mem == nullptr)
      {
        throw std::runtime_error("Failed to create ARKODE memory");
      }

      // Create the user-supplied function table to store the Python user functions
      auto cb_fns = arkode_user_supplied_fn_table_alloc();

      // Smuggle the user-supplied function table into callback wrappers through the user_data pointer
      int ark_status = ARKodeSetUserData(ark_mem, static_cast<void*>(cb_fns));
      if (ark_status != ARK_SUCCESS)
      {
        free(cb_fns);
        throw std::runtime_error("Failed to set user data in ARKODE memory");
      }

      // Ensure ARKodeFree will free the user-supplied function table
      ark_status = arkSetOwnUserData(ark_mem, SUNTRUE);
      if (ark_status != ARK_SUCCESS)
      {
        free(cb_fns);
        throw std::runtime_error(
          "Failed to set user data ownership in ARKODE memory");
      }

      // Finally, set the RHS functions
      cb_fns->arkstep_fe = nb::cast(fe);
      cb_fns->arkstep_fi = nb::cast(fi);

      return ark_mem;
    }, // .none() must be added to functions that accept nullptr as a valid argument
    nb::arg("fe").none(), nb::arg("fi").none(), nb::arg("t0"), nb::arg("y0"),
    nb::arg("sunctx"));

  m.def(
    "ARKStepCreateAdjointStepper",
    [](void* arkode_mem, std::function<std::remove_pointer_t<SUNAdjRhsFn>> adj_fe,
       std::function<std::remove_pointer_t<SUNAdjRhsFn>> adj_fi, sunrealtype tf,
       N_Vector sf, SUNContext sunctx) -> std::tuple<int, SUNAdjointStepper>
    {
      auto fe_wrapper = adj_fe ? arkstep_adjfe_wrapper : nullptr;
      auto fi_wrapper = adj_fi ? arkstep_adjfi_wrapper : nullptr;

      SUNAdjointStepper adj_stepper = nullptr;
      int ark_status = ARKStepCreateAdjointStepper(arkode_mem, fe_wrapper,
                                                   fi_wrapper, tf, sf, sunctx,
                                                   &adj_stepper);
      if (ark_status != ARK_SUCCESS)
      {
        throw std::runtime_error(
          "Failed to create adjoint stepper in py-sundials memory");
      }

      // Finally, set the RHS functions
      void* user_data = nullptr;
      ark_status      = ARKodeGetUserData(arkode_mem, &user_data);
      if (ark_status != ARK_SUCCESS)
      {
        throw std::runtime_error("Failed to extract ARKODE user data");
      }

      auto cb_fns = static_cast<arkode_user_supplied_fn_table*>(user_data);

      cb_fns->arkstep_adjfe = nb::cast(adj_fe);
      cb_fns->arkstep_adjfi = nb::cast(adj_fi);

      return std::make_tuple(ark_status, adj_stepper);
    },
    nb::arg("arkode_mem"), nb::arg("adj_fe").none(), nb::arg("adj_fi").none(),
    nb::arg("tf"), nb::arg("sf"), nb::arg("sunctx"));

  m.def(
    "ARKStepGetCurrentButcherTables",
    [](void* ark_mem)
    {
      ARKodeButcherTable fe = nullptr;
      ARKodeButcherTable fi = nullptr;

      int status = ARKStepGetCurrentButcherTables(ark_mem, &fi, &fe);

      return std::make_tuple(status, fi, fe);
    },
    "WARNING: this function returns ARKodeButcherTable references, DO NOT WRAP "
    "THEM IN A `ARKodeButcherTableView`. Doing so will result in a double free "
    "or worse.",
    nb::rv_policy::reference);
}

} // namespace pysundials
