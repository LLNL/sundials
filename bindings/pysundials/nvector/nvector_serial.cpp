#include <nanobind/nanobind.h>
#include <nanobind/ndarray.h>
#include <nanobind/stl/vector.h>

#include <nvector/nvector_serial.h>
#include <sundials/sundials_core.h>

#include "pysundials_types.hpp"

namespace nb = nanobind;

namespace pysundials {

void bind_nvector_serial(nb::module_& m)
{
#include "nvector_serial_generated.hpp"
}

} // namespace pysundials
