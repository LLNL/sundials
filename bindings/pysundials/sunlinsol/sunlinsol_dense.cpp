#include <nanobind/nanobind.h>

#include <sundials/sundials_core.hpp>
#include <sunlinsol/sunlinsol_dense.h>

namespace nb = nanobind;

void bind_sunlinsol_dense(nb::module_& m)
{
  m.def("SUNLinSol_Dense", &SUNLinSol_Dense, nb::rv_policy::reference);
}