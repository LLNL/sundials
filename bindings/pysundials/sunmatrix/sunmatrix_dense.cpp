/* -----------------------------------------------------------------
 * Programmer(s): Cody J. Balos @ LLNL
 * -----------------------------------------------------------------
 * SUNDIALS Copyright Start
 * Copyright (c) 2025-2025, Lawrence Livermore National Security,
 * University of Maryland Baltimore County, and the SUNDIALS contributors.
 * Copyright (c) 2013-2025, Lawrence Livermore National Security
 * and Southern Methodist University.
 * Copyright (c) 2002-2013, Lawrence Livermore National Security.
 * All rights reserved.
 *
 * See the top-level LICENSE and NOTICE files for details.
 *
 * SPDX-License-Identifier: BSD-3-Clause
 * SUNDIALS Copyright End
 * -----------------------------------------------------------------*/

#include <nanobind/nanobind.h>
#include <nanobind/ndarray.h>

#include <sundials/sundials_core.hpp>
#include <sunmatrix/sunmatrix_dense.h>

namespace nb = nanobind;

void bind_sunmatrix_dense(nb::module_& m)
{
#include "sunmatrix_dense_generated.hpp"

  m.def(
    "SUNDenseMatrix_Data",
    [](SUNMatrix A)
    {
      auto rows  = static_cast<size_t>(SUNDenseMatrix_Rows(A));
      auto cols  = static_cast<size_t>(SUNDenseMatrix_Columns(A));
      auto owner = nb::find(A);
      auto ptr   = SUNDenseMatrix_Data(A);
      // SUNDenseMatrix_Data returns a column-major ordered array (i.e., Fortran style)
      return nb::ndarray<sunrealtype, nb::numpy, nb::ndim<2>,
                         nb::f_contig>(ptr, {rows, cols}, owner);
    },
    nb::arg("A"), nb::rv_policy::reference);
}
