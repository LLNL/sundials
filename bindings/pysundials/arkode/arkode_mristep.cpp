/*------------------------------------------------------------------------------
 * Programmer(s): Cody J. Balos @ LLNL
 *------------------------------------------------------------------------------
 * SUNDIALS Copyright Start
 * Copyright (c) 2025-2025, Lawrence Livermore National Security,
 * University of Maryland Baltimore County, and the SUNDIALS contributors.
 * Copyright (c) 2013-2025, Lawrence Livermore National Security
 * and Southern Methodist University.
 * Copyright (c) 2002-2013, Lawrence Livermore National Security.
 * All rights reserved.
 *
 * See the top-level LICENSE and NOTICE files for details.
 *
 * SPDX-License-Identifier: BSD-3-Clause
 * SUNDIALS Copyright End
 *----------------------------------------------------------------------------*/

#include <nanobind/nanobind.h>
#include <nanobind/ndarray.h>
#include <nanobind/stl/function.h>
#include <nanobind/stl/tuple.h>

#include <arkode/arkode.hpp>
#include <arkode/arkode_mristep.hpp>
#include <sundials/sundials_core.hpp>

#include "arkode_mristep_impl.h"

#include "arkode_usersupplied.hpp"

namespace nb = nanobind;

using namespace sundials::experimental;

void bind_arkode_mristep(nb::module_& m)
{
#include "arkode_mristep_generated.hpp"

  nb::class_<MRIStepCouplingView>(m, "MRIStepCouplingView")
    .def_static("Create", &MRIStepCouplingView::Create<MRIStepCouplingView>)
    .def("get", nb::overload_cast<>(&MRIStepCouplingView::get, nb::const_),
         nb::rv_policy::reference);

  // _MRIStepInnerStepper is a opaque/private class forward declared in a public header but
  // defined in a source file elsewhere. As such, we need to declare it here since its
  // not picked up in any header files by the generator.
  nb::class_<_MRIStepInnerStepper>(m, "_MRIStepInnerStepper");

  nb::class_<MRIStepInnerStepperView>(m, "MRIStepInnerStepperView")
    .def_static("Create", &MRIStepInnerStepperView::Create<MRIStepInnerStepper>)
    .def("get", nb::overload_cast<>(&MRIStepInnerStepperView::get, nb::const_),
         nb::rv_policy::reference);

  m.def(
    "MRIStepInnerStepper_Create",
    [](SUNContext sunctx)
    {
      MRIStepInnerStepper stepper = nullptr;

      int status      = MRIStepInnerStepper_Create(sunctx, &stepper);
      auto cb_fns     = mristepinnerstepper_user_supplied_fn_table_alloc();
      stepper->python = static_cast<void*>(cb_fns);

      return std::make_tuple(status, stepper);
    },
    nb::rv_policy::reference);

  m.def(
    "MRIStepInnerStepper_CreateFromSUNStepper",
    [](SUNStepper stepper)
    {
      MRIStepInnerStepper inner_stepper = nullptr;

      int status = MRIStepInnerStepper_CreateFromSUNStepper(stepper,
                                                            &inner_stepper);

      return std::make_tuple(status, stepper);
    },
    nb::rv_policy::reference);

  m.def(
    "ARKodeCreateMRIStepInnerStepper",
    [](void* inner_arkode_mem)
    {
      MRIStepInnerStepper stepper = nullptr;

      int status = ARKodeCreateMRIStepInnerStepper(inner_arkode_mem, &stepper);

      return std::make_tuple(status, stepper);
    },
    nb::rv_policy::reference);

  m.def(
    "MRIStepCreate",
    [](std::function<std::remove_pointer_t<ARKRhsFn>> fse,
       std::function<std::remove_pointer_t<ARKRhsFn>> fsi, sunrealtype t0,
       N_Vector y0, MRIStepInnerStepper stepper, SUNContext sunctx)
    {
      auto fse_wrapper = fse ? mristep_fse_wrapper : nullptr;
      auto fsi_wrapper = fsi ? mristep_fsi_wrapper : nullptr;

      void* ark_mem = MRIStepCreate(fse_wrapper, fsi_wrapper, t0, y0, stepper,
                                    sunctx);
      if (ark_mem == nullptr)
      {
        throw std::runtime_error("Failed to create ARKODE memory");
      }

      // Create the user-supplied function table to store the Python user functions
      auto cb_fns = arkode_user_supplied_fn_table_alloc();

      // Smuggle the user-supplied function table into callback wrappers through the user_data pointer
      int ark_status = ARKodeSetUserData(ark_mem, static_cast<void*>(cb_fns));
      if (ark_status != ARK_SUCCESS)
      {
        free(cb_fns);
        throw std::runtime_error("Failed to set user data in ARKODE memory");
      }

      // Ensure ARKodeFree will free the user-supplied function table
      ark_status = arkSetOwnUserData(ark_mem, SUNTRUE);
      if (ark_status != ARK_SUCCESS)
      {
        free(cb_fns);
        throw std::runtime_error(
          "Failed to set user data ownership in ARKODE memory");
      }

      // Finally, set the RHS function
      cb_fns->mristep_fse = nb::cast(fse);
      cb_fns->mristep_fsi = nb::cast(fsi);

      return ark_mem;
    },
    nb::arg("fse").none(), nb::arg("fsi").none(), nb::arg("t0"), nb::arg("y0"),
    nb::arg("inner_stepper"), nb::arg("sunctx"));

  m.def(
    "MRIStepGetCurrentCoupling",
    [](void* ark_mem)
    {
      MRIStepCoupling C = nullptr;

      int status = MRIStepGetCurrentCoupling(ark_mem, &C);

      return std::make_tuple(status, C);
    },
    "WARNING: this function returns a MRIStepCoupling reference, DO NOT WRAP "
    "IT IN A `MRIStepCouplingView`. Doing so will result in a double free "
    "or worse.",
    nb::rv_policy::reference);
}
