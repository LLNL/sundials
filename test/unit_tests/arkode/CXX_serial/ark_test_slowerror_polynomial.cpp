/*-----------------------------------------------------------------
 * Programmer(s): Daniel R. Reynolds @ SMU
 *---------------------------------------------------------------
 * SUNDIALS Copyright Start
 * Copyright (c) 2002-2024, Lawrence Livermore National Security
 * and Southern Methodist University.
 * All rights reserved.
 *
 * See the top-level LICENSE and NOTICE files for details.
 *
 * SPDX-License-Identifier: BSD-3-Clause
 * SUNDIALS Copyright End
 *---------------------------------------------------------------
 * Routine to test an MRI method's embedding-based error
 * estimate.  Uses a simple polynomial test problem with 1
 * component,
 *    dy/dt = a*t^2 + b*t + c,  y(0) = 1.
 *
 * We run a single time step, using a variety of slow step sizes,
 * H = {1, 1/2, 1/4, 1/8, 1/16}.
 *
 * We place the entire ODE in the "slow" RHS partition.  All tests
 * use ARKODE's default fifth-order ERK method, with relative and
 * absolute tolerances set to 1e-10 and 1e-12, respectively.
 *
 * When using an IMEX method (e.g., IMEX-MRI-SR), the entire slow
 * partition is treated implicitly.
 *
 * We select the slow integrator based on a command-line argument,
 * with the default being ARKODE_MRI_GARK_ERK22a.
 *
 * The program should be run with arguments in the following order:
 *   $ a.out method a b c
 * Not all arguments are required, but these must be omitted from
 * end-to-beginning, i.e. any one of
 *   $ a.out method a b
 *   $ a.out method a
 *   $ a.out method
 *   $ a.out
 * are acceptable.  We require that the "method" argument be a
 * string corresponding to a valid embedded ARKODE_MRITableID.
 *-----------------------------------------------------------------*/

// Header files
#include <algorithm>
#include <arkode/arkode_arkstep.h>
#include <arkode/arkode_mristep.h>
#include <cmath>
#include <iostream>
#include <nvector/nvector_serial.h>
#include <stdio.h>
#include <string.h>
#include <sundials/sundials_core.hpp>
#include <sunlinsol/sunlinsol_dense.h>
#include <sunmatrix/sunmatrix_dense.h>
#include <vector>

#if defined(SUNDIALS_EXTENDED_PRECISION)
#define GSYM "Lg"
#define ESYM "Le"
#define FSYM "Lf"
#else
#define GSYM "g"
#define ESYM "e"
#define FSYM "f"
#endif

#define ZERO SUN_RCONST(0.0)
#define ONE  SUN_RCONST(1.0)
#define TWO  SUN_RCONST(2.0)

using namespace std;

// User data structure
struct UserData
{
  sunrealtype a;
  sunrealtype b;
  sunrealtype c;
};

// User-supplied Functions Called by the Solver
static int f0(sunrealtype t, N_Vector y, N_Vector ydot, void* user_data);
static int fn(sunrealtype t, N_Vector y, N_Vector ydot, void* user_data);
static int Jn(sunrealtype t, N_Vector y, N_Vector fy, SUNMatrix J,
              void* user_data, N_Vector tmp1, N_Vector tmp2, N_Vector tmp3);

// Private utility functions
static int Ytrue(sunrealtype t, N_Vector y, UserData& udata);
static int computeErrorWeights(N_Vector ycur, N_Vector weight, sunrealtype rtol,
                               sunrealtype atol, N_Vector vtemp);
static int check_retval(void* returnvalue, const char* funcname, int opt);
static int run_test(void* mristep_mem, N_Vector y, sunrealtype T0,
                    vector<sunrealtype>& Hvals, string method,
                    sunrealtype reltol, sunrealtype abstol, UserData& udata);

// Main Program
int main(int argc, char* argv[])
{
  // general problem parameters
  sunrealtype T0   = ZERO;                 // initial time
  sunindextype NEQ = 1;                    // number of dependent vars.
  string method;                           // MRI method name
  sunrealtype reltol = SUN_RCONST(1.e-10); // fast solver tolerances
  sunrealtype abstol = SUN_RCONST(1.e-12);

  // general problem variables
  int retval;     // reusable error-checking flag
  UserData udata; // user-data structure

  //
  // Initialization
  //

  // Retrieve the command-line options:  method Npart ep test
  if (argc > 1) { method = argv[1]; }
  else { method = "ARKODE_MRI_GARK_ERK22a"; }
  if (argc > 2) { udata.a = SUNStrToReal(argv[2]); }
  else { udata.a = ONE; }
  if (argc > 3) { udata.b = SUNStrToReal(argv[3]); }
  else { udata.b = ONE; }
  if (argc > 4) { udata.c = SUNStrToReal(argv[4]); }
  else { udata.c = ONE; }

  sunbooleantype implicit = SUNFALSE;
  if ((method == "ARKODE_MRI_GARK_IRK21a") ||
      (method == "ARKODE_MRI_GARK_ESDIRK34a") ||
      (method == "ARKODE_MRI_GARK_ESDIRK46a") ||
      (method == "ARKODE_IMEX_MRI_SR21") ||
      (method == "ARKODE_IMEX_MRI_SR32") || (method == "ARKODE_IMEX_MRI_SR43"))
  {
    implicit = SUNTRUE;
  }

  // Initial problem output (and set implicit solver tolerances as needed)
  cout << "\nSlow error estimation test (polynomial ODE problem):\n";
  cout << "    problem parameters:  a = " << udata.a << ",  b = " << udata.b
       << ",  c = " << udata.c << endl;
  cout << "    MRI method: " << method;
  if (implicit) { cout << " (implicit)" << endl; }
  else { cout << " (explicit)" << endl; }

  //
  // Problem Setup
  //

  // Create SUNDIALS context
  sundials::Context ctx;

  // Create serial vectors for the solution and reference
  N_Vector y = N_VNew_Serial(NEQ, ctx);
  if (check_retval((void*)y, "N_VNew_Serial", 0)) return 1;
  retval = Ytrue(T0, y, udata);
  if (check_retval(&retval, "Ytrue", 1)) return 1;

  // Set up fast ARKStep integrator as fifth-order adaptive ERK
  void* inner_arkode_mem = ARKStepCreate(f0, NULL, T0, y, ctx);
  if (check_retval((void*)inner_arkode_mem, "ARKStepCreate", 0)) return 1;
  retval = ARKodeSetOrder(inner_arkode_mem, 5);
  if (check_retval(&retval, "ARKodeSetOrder", 1)) return 1;
  retval = ARKodeSStolerances(inner_arkode_mem, reltol, abstol);
  if (check_retval(&retval, "ARKodeSStolerances", 1)) return 1;
  retval = ARKodeSetMaxNumSteps(inner_arkode_mem, 1000000);
  if (check_retval(&retval, "ARKodeSetMaxNumSteps", 1)) return (1);

  // Create inner stepper wrapper
  MRIStepInnerStepper inner_stepper = NULL; // inner stepper
  retval = ARKodeCreateMRIStepInnerStepper(inner_arkode_mem, &inner_stepper);
  if (check_retval(&retval, "ARKodeCreateMRIStepInnerStepper", 1)) return 1;

  // Set up slow MRIStep integrator
  void* mristep_mem = NULL;
  if (implicit)
  {
    mristep_mem = MRIStepCreate(NULL, fn, T0, y, inner_stepper, ctx);
  }
  else { mristep_mem = MRIStepCreate(fn, NULL, T0, y, inner_stepper, ctx); }
  if (check_retval((void*)mristep_mem, "MRIStepCreate", 0)) return 1;
  MRIStepCoupling C = MRIStepCoupling_LoadTableByName(method.c_str());
  if (check_retval((void*)C, "MRIStepCoupling_LoadTableByName", 0)) return 1;
  retval = MRIStepSetCoupling(mristep_mem, C);
  if (check_retval(&retval, "MRIStepSetCoupling", 1)) return 1;
  SUNMatrix A        = NULL; // matrix for slow solver
  SUNLinearSolver LS = NULL; // slow linear solver object
  if (implicit)
  {
    A = SUNDenseMatrix(NEQ, NEQ, ctx);
    if (check_retval((void*)A, "SUNDenseMatrix", 0)) return 1;
    LS = SUNLinSol_Dense(y, A, ctx);
    if (check_retval((void*)LS, "SUNLinSol_Dense", 0)) return 1;
    retval = ARKodeSetLinearSolver(mristep_mem, LS, A);
    if (check_retval(&retval, "ARKodeSetLinearSolver", 1)) return 1;
    retval = ARKodeSetJacFn(mristep_mem, Jn);
    if (check_retval(&retval, "ARKodeSetJacFn", 1)) return 1;
    retval = ARKodeSetJacEvalFrequency(mristep_mem, 1);
    if (check_retval(&retval, "ARKodeSetJacEvalFrequency", 1)) return 1;
    retval = ARKodeSetLSetupFrequency(mristep_mem, 1);
    if (check_retval(&retval, "ARKodeSetLSetupFrequency", 1)) return 1;
    retval = ARKodeSetMaxNonlinIters(mristep_mem, 20);
    if (check_retval(&retval, "ARKodeSetMaxNonlinIters", 1)) return 1;
  }
  retval = ARKodeSStolerances(mristep_mem, reltol, abstol);
  if (check_retval(&retval, "ARKodeSStolerances", 1)) return 1;
  retval = ARKodeSetUserData(mristep_mem, (void*)&udata);
  if (check_retval(&retval, "ARKodeSetUserData", 1)) return 1;
  retval = ARKodeSetAccumulatedErrorType(mristep_mem, ARK_ACCUMERROR_MAX);
  if (check_retval(&retval, "ARKodeSetAccumulatedErrorType", 1)) return 1;

  // Run test for various H values
  vector<sunrealtype> Hvals(5);
  for (size_t i = 0; i < Hvals.size(); i++)
  {
    Hvals[i] = SUN_RCONST(0.01) / SUNRpowerI(SUN_RCONST(2.0), (int)i);
  }
  retval = run_test(mristep_mem, y, T0, Hvals, method, reltol, abstol, udata);
  if (check_retval(&retval, "run_test", 1)) return 1;

  // Clean up and return
  MRIStepCoupling_Free(C);
  ARKodeFree(&inner_arkode_mem);
  MRIStepInnerStepper_Free(&inner_stepper);
  ARKodeFree(&mristep_mem);
  if (LS) { SUNLinSolFree(LS); } // free system linear solver
  if (A) { SUNMatDestroy(A); }   // free system matrix
  N_VDestroy(y);                 // Free y and yref vectors
  return 0;
}

//------------------------------
// Functions called by the solver
//------------------------------

static int f0(sunrealtype t, N_Vector y, N_Vector ydot, void* user_data)
{
  // fill in the RHS function with zeros and return with success
  N_VConst(ZERO, ydot);
  return 0;
}

static int fn(sunrealtype t, N_Vector y, N_Vector ydot, void* user_data)
{
  // fill in the RHS function and return with success
  UserData* udata     = (UserData*)user_data;
  sunrealtype* dydata = N_VGetArrayPointer(ydot);
  dydata[0]           = udata->a * t * t + udata->b * t + udata->c;
  return 0;
}

static int Jn(sunrealtype t, N_Vector y, N_Vector fy, SUNMatrix J,
              void* user_data, N_Vector tmp1, N_Vector tmp2, N_Vector tmp3)
{
  // fill in the Jacobian and return with success
  SM_ELEMENT_D(J, 0, 0) = ZERO;
  return 0;
}

//------------------------------
// Private helper functions
//------------------------------

static int run_test(void* mristep_mem, N_Vector y, sunrealtype T0,
                    vector<sunrealtype>& Hvals, string method,
                    sunrealtype reltol, sunrealtype abstol, UserData& udata)
{
  // Reused variables
  int retval;
  sunrealtype t;
  N_Vector vtemp     = N_VClone(y);
  N_Vector ele       = N_VClone(y);
  N_Vector ewt       = N_VClone(y);
  sunrealtype* ydata = N_VGetArrayPointer(y);
  sunrealtype* vdata = N_VGetArrayPointer(vtemp);

  // Set storage for errors
  vector<sunrealtype> dsm(Hvals.size(), ZERO);
  vector<sunrealtype> dsm_est(Hvals.size(), ZERO);

  // Loop over step sizes
  for (size_t iH = 0; iH < Hvals.size(); iH++)
  {
    // Reset integrator for this run
    t      = T0;
    retval = Ytrue(t, y, udata);
    if (check_retval(&retval, "Ytrue", 1)) return 1;
    retval = ARKodeReset(mristep_mem, t, y);
    if (check_retval(&retval, "ARKodeReset", 1)) return 1;
    retval = ARKodeSetFixedStep(mristep_mem, Hvals[iH]);
    if (check_retval(&retval, "ARKodeSetFixedStep", 1)) return 1;
    retval = ARKodeResetAccumulatedError(mristep_mem);
    if (check_retval(&retval, "ARKodeResetAccumulatedError", 1)) return 1;

    // Run ARKodeEvolve to compute one step
    retval = ARKodeEvolve(mristep_mem, t + Hvals[iH], y, &t, ARK_ONE_STEP);
    if (check_retval(&retval, "ARKodeEvolve", 1)) return 1;
    retval = ARKodeGetEstLocalErrors(mristep_mem, ele);
    if (check_retval(&retval, "ARKodeGetEstLocalErrors", 1)) return 1;
    retval = computeErrorWeights(y, ewt, reltol, abstol, vtemp);
    if (check_retval(&retval, "computeErrorWeights", 1)) return 1;
    dsm_est[iH] = N_VWrmsNorm(ewt, ele);

    // Compute/print solution error
    retval = Ytrue(t, vtemp, udata);
    if (check_retval(&retval, "Ytrue", 1)) return 1;
    dsm[iH] = abs(ydata[0] - vdata[0]) / (abstol + reltol * abs(vdata[0]));
    if (method == "ARKODE_MRI_GARK_ERK22a")
    {
      printf("       H  %.5f    dsm  %.2e    dsm_est  %.2e    dsm_anal  %.2e   "
             " dsm_est_anal  %.2e\n",
             Hvals[iH], dsm[iH], dsm_est[iH],
             Hvals[iH] * Hvals[iH] * Hvals[iH] * abs(udata.a / 12.0) /
               (abstol + reltol * abs(vdata[0])),
             Hvals[iH] * Hvals[iH] *
               abs(udata.a * Hvals[iH] / 4.0 + udata.b / 2.0) /
               (abstol + reltol * abs(vdata[0])));
    }
    else if (method == "ARKODE_MRI_GARK_IRK21a")
    {
      printf("       H  %.5f    dsm  %.2e    dsm_est  %.2e    dsm_anal  %.2e   "
             " dsm_est_anal  %.2e\n",
             Hvals[iH], dsm[iH], dsm_est[iH],
             Hvals[iH] * Hvals[iH] * Hvals[iH] * abs(udata.a / 6.0) /
               (abstol + reltol * abs(vdata[0])),
             Hvals[iH] * Hvals[iH] *
               abs(udata.a * Hvals[iH] / 2.0 + udata.b / 2.0) /
               (abstol + reltol * abs(vdata[0])));
    }
    else if (method == "ARKODE_IMEX_MRI_SR21")
    {
      printf("       H  %.5f    dsm  %.2e    dsm_est  %.2e    dsm_anal  %.2e   "
             " dsm_est_anal  %.2e\n",
             Hvals[iH], dsm[iH], dsm_est[iH],
             Hvals[iH] * Hvals[iH] * Hvals[iH] * abs(udata.a * 3137.0 / 50370.0) /
               (abstol + reltol * abs(vdata[0])),
             Hvals[iH] * Hvals[iH] *
               abs(udata.a * Hvals[iH] * 20191.0 / 755550.0 -
                   udata.b * 19.0 / 30.0) /
               (abstol + reltol * abs(vdata[0])));
    }
    else
    {
      printf("       H  %.5f    dsm  %.2e    dsm_est  %.2e\n", Hvals[iH],
             dsm[iH], dsm_est[iH]);
    }
  }

  N_VDestroy(ele);
  N_VDestroy(ewt);
  N_VDestroy(vtemp);
  return (0);
}

static int Ytrue(sunrealtype t, N_Vector y, UserData& udata)
{
  sunrealtype* ydata = N_VGetArrayPointer(y);
  ydata[0]           = udata.a / SUN_RCONST(3.0) * t * t * t +
             udata.b / SUN_RCONST(2.0) * t * t + udata.c * t + ONE;
  return (0);
}

/* Error weight calculation routine (mimics what's in ARKODE already) */
static int computeErrorWeights(N_Vector ycur, N_Vector weight, sunrealtype rtol,
                               sunrealtype atol, N_Vector vtemp)
{
  N_VAbs(ycur, vtemp);
  N_VScale(rtol, vtemp, vtemp);
  N_VAddConst(vtemp, atol, vtemp);
  N_VInv(vtemp, weight);
  return (0);
}

/* Check function return value...
    opt == 0 means SUNDIALS function allocates memory so check if
             returned NULL pointer
    opt == 1 means SUNDIALS function returns a retval so check if
             retval >= 0
    opt == 2 means function allocates memory so check if returned
             NULL pointer
*/
static int check_retval(void* returnvalue, const char* funcname, int opt)
{
  int* retval;

  // Check if SUNDIALS function returned NULL pointer - no memory allocated
  if (opt == 0 && returnvalue == NULL)
  {
    fprintf(stderr, "\nSUNDIALS_ERROR: %s() failed - returned NULL pointer\n\n",
            funcname);
    return 1;
  }

  // Check if retval < 0
  else if (opt == 1)
  {
    retval = (int*)returnvalue;
    if (*retval < 0)
    {
      fprintf(stderr, "\nSUNDIALS_ERROR: %s() failed with retval = %d\n\n",
              funcname, *retval);
      return 1;
    }
  }

  // Check if function returned NULL pointer - no memory allocated
  else if (opt == 2 && returnvalue == NULL)
  {
    fprintf(stderr, "\nMEMORY_ERROR: %s() failed - returned NULL pointer\n\n",
            funcname);
    return 1;
  }

  return 0;
}

/*---- end of file ----*/
