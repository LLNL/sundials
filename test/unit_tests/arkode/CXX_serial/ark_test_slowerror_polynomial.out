
Slow error estimation test (polynomial ODE problem):
    problem parameters:  a = 1,  b = 1,  c = 1
    MRI method: ARKODE_MRI_GARK_ERK22a (explicit)
       H  0.01000    dsm  8.17e+02    dsm_est  4.93e+05    dsm_anal  8.17e+02    dsm_est_anal  4.93e+05
       H  0.00500    dsm  1.03e+02    dsm_est  1.23e+05    dsm_anal  1.03e+02    dsm_est_anal  1.23e+05
       H  0.00250    dsm  1.29e+01    dsm_est  3.09e+04    dsm_anal  1.29e+01    dsm_est_anal  3.09e+04
       H  0.00125    dsm  1.61e+00    dsm_est  7.73e+03    dsm_anal  1.61e+00    dsm_est_anal  7.73e+03
       H  0.00063    dsm  2.01e-01    dsm_est  1.93e+03    dsm_anal  2.01e-01    dsm_est_anal  1.93e+03
